lesson21/main.js

```diff
@@ -118,24 +121,100 @@ const formingTableHeaderNameWithKey = (key) => {
 
 const createTableData = (usersData) => {
   const fragment = document.createDocumentFragment();
-  usersData.forEach((userData) => {
-    const trOfTdata = document.createElement("tr");
+  usersData.forEach((userData, index) => {
+    const trOfTdata = createElementWithClassName("tr", "users-table__tr-td");
+    const usersIndex = ++index;
     /**
      *Extract value from each of usersData
      * @param {Object} userData the object of each users data
      * @param {String,Number} val value of each users data
      */
 
-    Object.values(userData).forEach((val) => {
+    Object.keys(userData).forEach((key) => {
       const td = createElementWithClassName("td", "users-table__td");
-      td.textContent = val;
+      key === "id"
+        ? (td.textContent = usersIndex)
+        : (td.textContent = userData[key]);
       trOfTdata.appendChild(td);
     });
     fragment.appendChild(trOfTdata);
   });
   return fragment;
 };
 
+const createSortBtn = () => {
+  const btn = createElementWithClassName("button", "sort-btn");
+  const sortArrow = createElementWithClassName("img", "sort-img");
+  // bothのイメージをデフォルトとする
+  sortArrow.src = "./img/both.svg";
+  btn.appendChild(sortArrow);
+  return btn;
+};
+
+const renderSortBtn = () => {
+  const btn = createSortBtn();
+  const targetEle = document.getElementById("id");
+  targetEle.appendChild(btn);
+};
+
+const clickSortBtn = () => {
+  const sortArrow = document.querySelector(".sort-btn");
+  sortArrow.addEventListener("click", () => {
+    changeSortStateAndArrowImg(sortArrow);
+    rerenderTableData(newUsersData);
+  });
+};
+
+const getNewUsersData = () => {
+  const trOfTdata = document.querySelectorAll(".users-table__tr-td");
+  return (newUsersData = Array.from(trOfTdata));
+};
+
+const changeSortStateAndArrowImg = (sortArrow) => {
+  if (sortState === "both") {
+    sortState = "asc";
+    sortArrow.src = "/img/asc.svg";
+  } else if (sortState === "asc") {
+    sortState = "desc";
+    sortArrow.src = "/img/desc.svg";
+  } else {
+    sortState = "both";
+    sortArrow.src = "/img/both.svg";
+  }
+};
+
+const rerenderTableData = (newUsersData) => {
+  const table = document.getElementById("js-table");
+  const th = document.getElementById("id");
+  const colNum = th.cellIndex;
+  if (sortState === "both") {
+    sortInit(newUsersData);
+  } else if (sortState === "asc") {
+    sortAsc(newUsersData, colNum);
+  } else {
+    sortDesc(newUsersData, colNum);
+  }
+  table.append(...newUsersData);
+};
+
+const sortAsc = (usersData, colNum) => {
+  usersData.sort((a, b) => {
+    return a.cells[colNum].innerHTML - b.cells[colNum].innerHTML;
+  });
+};
+
+const sortDesc = (usersData, colNum) => {
+  usersData.sort((a, b) => {
+    return b.cells[colNum].innerHTML - a.cells[colNum].innerHTML;
+  });
+};
+
+const sortInit = (usersData) => {
+  usersData.sort(() => {
+    return 0.5 - Math.random();
+  });
+};
```

Comment:
MUST
I quote Mr. Morita's comment
> dataを受け取ってそのまま破壊的なメソッドを使ってはいけないです

Reply:
You reminded me `sort()` is the method which changes the original array.
I implemented it with `copiedUsersData( = [...usersData])` here.
Thanks!
[84b7989](https://github.com/yuka830/js-lessons/pull/27/commits/84b798958516a9c5dc751239ed0786f978488832)